### DISCUSSION TITLE
authenticated user form with multple writes

### DISCUSSION THREAD

**Discussion by St897** [sqlpage/SQLPage#259] [Q&A]
Hello,

i'm trying to setup a insert form which only works for an authenticated user

form.sql
 - check if user is logged in
 - redirect user if not
 - general shell for the webpage
 - form (select) with action

action.sql
 - check if user is logged in
 - redirect user if not
 - general shell for the webpage
 - start transaction
 - insert into table 1
 - insert into table 2
 - commit;
 - redirect

i would like to add a redirect in the action.sql returning to the form.sql (with or without the filled in data, so it can be edited)

but i always get this error, 

_The redirect component cannot be used after data has already been sent to the client's browser. This component must be used before any other component. To fix this, either move the call to the 'redirect' component to the top of the SQL file, or create a new SQL file where 'redirect' is the first component_

the suggestion seems clear, but i don't how see how to setup in that case the flow described?

(the goal is to have a form where only validated users can input with a sensible redirect,
where the underlying input goes to more than 1 table)

(inspired by:
* https://github.com/lovasoa/SQLpage/tree/main/examples/user-authentication
* https://github.com/lovasoa/SQLpage/blob/main/examples/modeling%20a%20many%20to%20many%20relationship%20with%20a%20form/write_submit.sql 
* and with a shell)

Any suggestions?

**Comment by lovasoa** [comment]
Hello and welcome to SQLPage ! The logic you mention in your description above seems correct to me. Can you share a small sql snippet that reproduces your issue ?

Are you sure the *check if user is logged in* step does not return anything to the browser ?

**Reply by St897** [reply]
i made a basic example on:

https://github.com/St897/SQLpage/tree/main/examples/auth_multiple_writes

it uses postgres as a backend;
with a start transaction and commit

**Reply by lovasoa** [reply]
In https://github.com/St897/SQLpage/blob/main/examples%2Fauth_multiple_writes%2Faction.sql you use the redirect component after having sent a shell.

**Reply by lovasoa** [reply]
If you want to redirect, then just remove the `select 'shell'  ...`
